### PR TITLE
Add missing_docs lint to the rustdoc lint group

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Just realized that the `rustdoc` lint group didn't have the `missing_docs` lint inside.

r? @jyn514

### CODE REVIEW

**Comment by jyn514** [comment]
I don't think I'm the right person to review this.

r? @ollie27
cc @rust-lang/lang

**Review by pickfire — APPROVED** [review]
Looks good to me. (don't take/rollup based on this review)

**Comment by ollie27** [comment]
When the `rustdoc` lint group was added in #56689 `missing_docs` was deliberately left out. Adding it would be changing what the group is for.

**Comment by GuillaumeGomez** [comment]
The lint is emitted by rustdoc too. :)

**Comment by ollie27** [comment]
> The lint is emitted by rustdoc too. :)

Yes, but it's also emitted by rustc and the `rustdoc` lint group is for lints only emitted by rustdoc. From #56689:

> I deliberately didn't include `missing_docs` because this is kind of a stepping stone for moving our lints into tool lints (i.e. `#![warn(rustdoc::private_doc_tests)]`), since all of these are specifically emitted by rustdoc.

My main question is what would the `rustdoc` lint group be for if `missing_docs` is added?

**Comment by GuillaumeGomez** [comment]
Literally what it's supposed to be: lints emitted by rustdoc (even if they are also emitted by/from rustc).

**Comment by jyn514** [comment]
Rustdoc also emits lints about unknown and renamed lints (https://github.com/rust-lang/rust/pull/75903). But I think it's clear those shouldn't be part of the rustdoc group.

**Comment by GuillaumeGomez** [comment]
Good point. Then let me add more precision: "lints emitted by rustdoc (even if they are also emitted by/from rustc) and linked to documentation."

**Comment by camelid** [comment]
@GuillaumeGomez Could you resolve the merge conflicts?

**Comment by jyn514** [comment]
@camelid there's no point resolving the merge conflicts until the rustdoc team decides this is a change they want to accept (and IMO this should have feedback from T-compiler too).

**Comment by camelid** [comment]
Sorry, drive-by triaging 😄

**Comment by jyn514** [comment]
The discussion has kind of stalled out ... there are two viewpoints of what `warn(rustdoc)` means:

1. Lints specific to rustdoc: https://github.com/rust-lang/rust/pull/77364#issuecomment-703942992
2. Lints specific to documentation: https://github.com/rust-lang/rust/pull/77364#issuecomment-704257690

I don't think we can do anything here until we decide which of those `warn(rustdoc)` should be. Maybe someone from T-lang has opinions?

**Comment by nikomatsakis** [comment]
My intuition is that users will expect `rustdoc` to mean "documentation-related lints emitted by rustdoc", and that it doesn't matter so much whether the lint is emitted by rustc or not, but I think it'd be useful to dig just a bit more into the implications of each choice, because I'm not sure I fully understand them.

I think that the argument in favor of a more limited definition is something like this:

* If we say that `rustdoc` includes `missing_docs`, then when you do `#![warn(rustdoc)]`, rustc will start to emit warnings for missing docs -- and that might be confusing.

In other words, maybe it's surprising the `warn(rustdoc)` changes rustc behavior at all?

**Comment by jyn514** [comment]
Here's another thought: I want to eventually change rustdoc's lints to have a tool prefix (`warn(rustdoc::broken_intra_doc_links)`). In a world where all lints except `missing_docs` and `unknown_and_renamed_lints` have a `rustdoc::` prefix, does it still make sense for `warn(rustdoc)` to warn about missing_docs?

**Comment by camelid** [comment]
> Here's another thought: I want to eventually change rustdoc's lints to have a tool prefix (`warn(rustdoc::broken_intra_doc_links)`). In a world where all lints except `missing_docs` and `unknown_and_renamed_lints` have a `rustdoc::` prefix, does it still make sense for `warn(rustdoc)` to warn about missing_docs?

Won't it be confusing for people that most docs-related lints start with `rustdoc::`, *except* for `missing_docs`? It seems confusing to me.

**Comment by jyn514** [comment]
@camelid `missing_docs` is the only documentation-related lint that runs in rustc directly instead of rustdoc.

So maybe the better fix here is to move missing_docs *out* of rustc and into rustdoc instead? And then everything is consistent.

**Comment by GuillaumeGomez** [comment]
I'd prefer to have it running in rustdoc directly indeed. :)

I can make a PR for that if you want?

**Comment by camelid** [comment]
> @camelid `missing_docs` is the only documentation-related lint that runs in rustc directly instead of rustdoc.

I figured, but my point is that that's an implementation detail and not that relevant to the user.

> So maybe the better fix here is to move missing_docs _out_ of rustc and into rustdoc instead? And then everything is consistent.

The one thing is that then you won't be able to run `missing_docs` by running rustc; you'll have to run rustdoc instead. It's probably not an issue, though, because there's not much reason to have documentation if you don't build it :)

**Comment by jyn514** [comment]
> The one thing is that then you won't be able to run missing_docs by running rustc; you'll have to run rustdoc instead

That's exactly why I don't consider this an implementation detail.

> I can make a PR for that if you want?

I think this should have input from T-lang, I don't think it's really rustdoc's decision to make.

**Comment by GuillaumeGomez** [comment]
My proposition still stands. :p

**Comment by LeSeulArtichaut** [comment]
If `missing_docs` gets moved to rustdoc, would that mean that library/project maintainers that use `#[deny(missing_docs)]` won't get an error on `cargo check`?

**Comment by LeSeulArtichaut** [comment]
> It's probably not an issue, though, because there's not much reason to have documentation if you don't build it :)

But do people run rustdoc in CI, for example?

**Comment by camelid** [comment]
> > It's probably not an issue, though, because there's not much reason to have documentation if you don't build it :)
> 
> But do people run rustdoc in CI, for example?

Yes, but this would mean that you can't run `rustc --deny missing_docs`, you instead run `rustdoc --deny missing_docs`. I'm not fully sure what your question is :)

**Comment by GuillaumeGomez** [comment]
I wanted `cargo check` to run `rustdoc --check` too. The PR is [open](https://github.com/rust-lang/cargo/pull/8859) but I don't think it's gonna get merged any time soon...

**Comment by GuillaumeGomez** [comment]
Rustdoc lints are now declared in rustdoc so no need for this PR anymore.
